### ISSUE TITLE
gogs server key

### ISSUE THREAD

**Issue by waldemara** [gogs/gogs#7064]
### Discussed in https://github.com/gogs/gogs/discussions/7063

<div type='discussions-op-text'>

<sup>Originally posted by **waldemara** June 15, 2022</sup>
Recently, I am getting "Unable to negotiate with x.x.x.x port 22: no matching host key type found. Their offer: ssh-rsa" when connecting to gogs server. This can be temporarily fixed by adding 'HostkeyAlgorithms +ssh-rsa' to ssh config file but real problem is that gogs is using obsolete keys.
How do I change gogs server keys?  This is for the embedded ssh server.</div>

**Comment by waldemara** [comment]
As mentioned in the thread, an upgrade to 0.12.9 does not upgrade the keys. Instead it fails. So it knows the keys are old but takes no action and leaves the admin with no means to upgrade them manually. It also leaves the server dead.

**Comment by waldemara** [comment]
I managed to get the server to work again but I don't know what cleared the error. Perhaps it was some transient state. Here is the documentation of everything I did.
(1) The 'before'. The server was running gogs 0.12.3 on Alma 8 with selinux enforcing. Access via git from a Fedora 34 machine worked fine.
(2) Fedora update. I decided to upgrade the client machine from Fedora 34 to Fedora 36. After the upgrade, git pull was returning 'Unable to negotiate with x.x.x.x port 22: no matching host key type found. Their offer: ssh-rsa'.
(3) Temporary workaround. Googling the error, I found a work around: https://serverfault.com/questions/1092998/ssh-no-matching-host-key-type-found, which amounted to setting 'HostkeyAlgorithms +ssh-rsa' option in the ssh config file.
(4) Attempting to upgrade host keys on the gogs internal ssh server. I found out that the internal gogs server only offers rsa keys with no way to add any other key types.
(5) Attempting to upgrade gogs server. Realizing there is a new release, I decided to upgrade the server with hopes of the upgrade resolving the problem. I upgraded gogs to 0.12.9 but it did not solve anything. Worse, the server started with `[gogs.io/gogs/internal/ssh/ssh.go:151 func1()] SSH: Error on handshaking: ssh: no common algorithm for host key; client offered: [rsa-sha2-512-cert-v01@openssh.com rsa-sha2-256-cert-v01@openssh.com rsa-sha2-512 rsa-sha2-256 ssh-ed25519-cert-v01@openssh.com ecdsa-sha2-nistp256-cert-v01@openssh.com ecdsa-sha2-nistp384-cert-v01@openssh.com ecdsa-sha2-nistp521-cert-v01@openssh.com sk-ssh-ed25519-cert-v01@openssh.com sk-ecdsa-sha2-nistp256-cert-v01@openssh.com ssh-ed25519 ecdsa-sha2-nistp256 ecdsa-sha2-nistp384 ecdsa-sha2-nistp521 sk-ssh-ed25519@openssh.com sk-ecdsa-sha2-nistp256@openssh.com], server offered: [ssh-rsa]` and was not usable after that.
(6) Reverting back to 0.12.3. I reverted back to 0.12.3 but surprisingly, the server started with the same error as above.
(7) Hacking attempts. Reading source code in internal/ssh/ssh.go, it appears that the server would recreate keys if they're not found on the server (line 182). That was my chance of forcing the server to update keys. I removed ssh directory by renaming it and was expecting the server to re-create it. It did not happen. The server did not even start. I found out later this was due to selinux blocking it. But why selinux allowed to start the server before and decided to block it now, I will never know.
(8) Restoring upgrade. After eventually realizing that selinux started blocking gogs, I restored ssh directory, then re-applied bin_t label on the gogs binary. Amazingly, the server started without problems: no complaints about ssh keys. With this, I re-upgraded to 0.12.9, made sure bin_t label is set on gogs executable, and started the server. The upgraded server started fine without complaints about the ssh keys.
(9) Reverting workaround. With the server not complaining about the keys any longer, I removed 'HostkeyAlgorithms +ssh-rsa' option in the client machine (fedora 36) ssh config file. Amazingly, git access started working fine. Workaround was no longer needed. Why this is so, I will never know. The keys haven't changed, the code between 0.12.3 and 0.12.9 does not appear to change anything in key handling.
